### PR TITLE
Add Grant support for Catalog buckets and objects

### PR DESCRIPTION
**Pull request by marounkoussaifi**
The following features were added in this PR:

1. DB Entities for Buckets' and Objects' grants
2. REST controllers and services for all grants mainly to do: create, edit, delete and get
3. Updated the existing catalog controllers and services to handle the new grant features
4. Added functional tests for the Grant controllers and services

The following tasks remains to develop while I'm handling this PR reviews:

1. Bug fix regarding the number of objects in grant buckets
2. Fix the current integration tests
3. Write new integration tests to handle the new grant features

### CODE REVIEW

File: src/main/java/org/ow2/proactive/catalog/dto/CatalogObjectGrantMetaData.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+import java.util.Objects;
+
+import org.ow2.proactive.catalog.repository.entity.CatalogObjectGrantEntity;
+import org.springframework.hateoas.ResourceSupport;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+
+public class CatalogObjectGrantMetaData extends ResourceSupport {
```
Comment (aminelouati) [inline src/main/java/org/ow2/proactive/catalog/dto/CatalogObjectGrantMetaData.java:36]: CatalogObjectGrantMetaData -> CatalogObjectGrantMetadata

File: src/main/java/org/ow2/proactive/catalog/dto/BucketGrantMetadata.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+import java.util.Objects;
+
+import org.ow2.proactive.catalog.repository.entity.BucketGrantEntity;
+import org.springframework.hateoas.ResourceSupport;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+
+public class BucketGrantMetadata extends ResourceSupport {
```
Comment (aminelouati) [inline src/main/java/org/ow2/proactive/catalog/dto/BucketGrantMetadata.java:36]: I think that you can improve and shorten you code by using the following annotation
@Data 
@Getter
@EqualsAndHashCode

File: src/main/java/org/ow2/proactive/catalog/dto/CatalogObjectGrantMetaData.java (hunk 1 above)
Comment (aminelouati) [reply src/main/java/org/ow2/proactive/catalog/dto/CatalogObjectGrantMetaData.java:36]: Same remark as above

File: src/main/java/org/ow2/proactive/catalog/repository/BucketGrantRepository.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    @QueryHints({ @QueryHint(name = "javax.persistence.lock.timeout", value = "5000") })
+    @Query(value = "SELECT bge FROM BucketGrantEntity bge WHERE bge.profiteer = ?1 And bge.grantee='user'")
+    List<BucketGrantEntity> findAllGrantsAssignedToAUsername(String username);
+
+    @Lock(LockModeType.PESSIMISTIC_WRITE)
+    @QueryHints({ @QueryHint(name = "javax.persistence.lock.timeout", value = "5000") })
+    @Query(value = "SELECT bge FROM BucketGrantEntity bge WHERE bge.profiteer = ?1 And bge.grantee='group'")
+    List<BucketGrantEntity> findAllGrantsAssignedToAUserGroup(String userGroup);
+
+}
```
Comment (aminelouati) [inline src/main/java/org/ow2/proactive/catalog/repository/BucketGrantRepository.java:71]: I think that you can shorten and improve the readability of all the jpaRepo classes using JPA query method as explained here: https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#jpa.query-methods.query-creation

File: src/main/java/org/ow2/proactive/catalog/repository/CatalogObjectGrantRepository.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    @QueryHints({ @QueryHint(name = "javax.persistence.lock.timeout", value = "5000") })
+    @Query(value = "SELECT coge.bucketEntity.id FROM CatalogObjectGrantEntity coge WHERE coge.profiteer = ?1 AND coge.grantee='group'")
+    List<Long> findAllBucketsIdFromCatalogObjectGrantsAssignedToAUserGroup(String userGroup);
+
+    @Lock(LockModeType.PESSIMISTIC_WRITE)
+    @QueryHints({ @QueryHint(name = "javax.persistence.lock.timeout", value = "5000") })
+    @Query(value = "SELECT coge FROM CatalogObjectGrantEntity coge WHERE coge.bucketEntity.id=?1")
+    List<CatalogObjectGrantEntity> findAllCatalogObjectGrantsAssignedToABucket(long bucketId);
+
+}
```
Comment (aminelouati) [inline src/main/java/org/ow2/proactive/catalog/repository/CatalogObjectGrantRepository.java:89]: same remark as above

File: src/main/java/org/ow2/proactive/catalog/repository/entity/BucketGrantEntity.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "GRANT_SEQUENCE")
+    @GenericGenerator(name = "GRANT_SEQUENCE", strategy = "org.hibernate.id.enhanced.SequenceStyleGenerator", parameters = { @org.hibernate.annotations.Parameter(name = "sequence_name", value = "GRANT_SEQUENCE"),
+                                                                                                                             @org.hibernate.annotations.Parameter(name = "initial_value", value = "1"),
+                                                                                                                             @org.hibernate.annotations.Parameter(name = "increment_size", value = "1") })
+    @Column(name = "ID")
+    protected Long id;
+
+    // Type of this grant: User or Group
+    @Column(name = "GRANTEE", nullable = false)
```
Comment (aminelouati) [inline src/main/java/org/ow2/proactive/catalog/repository/entity/BucketGrantEntity.java:58]: // Type of this grant: User or Group -> // Type of this grant: user or group

File: src/main/java/org/ow2/proactive/catalog/repository/entity/BucketGrantEntity.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    protected String creator;
+
+    // Username or Group Name
+    @Column(name = "PROFITEER", nullable = false)
+    protected String profiteer;
+
+    // Access type: Admin, write or read
+    @Column(name = "ACCESS_TYPE", nullable = false)
+    protected String accessType;
+
```
Comment (aminelouati) [inline src/main/java/org/ow2/proactive/catalog/repository/entity/BucketGrantEntity.java:72]: I think that you have to use enum type for all these fields to control possible values

File: src/main/java/org/ow2/proactive/catalog/util/AccessType.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * If needed, contact us to obtain a release under GPL Version 2 or 3
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive.catalog.util;
+
+// TODO Transform into Enum
+public class AccessType {
```
Comment (aminelouati) [inline src/main/java/org/ow2/proactive/catalog/util/AccessType.java:29]: Please transform that into Enum

File: src/main/java/org/ow2/proactive/catalog/repository/entity/BucketGrantEntity.java (hunk 6 above)
Comment (marounkoussaifi) [reply src/main/java/org/ow2/proactive/catalog/repository/entity/BucketGrantEntity.java:72]: I already planned to change this one to an Enum once I work on the negative grants

**Comment by marounkoussaifi** [comment]
@aminelouati I have solved all of your comments except one. The access type of a grant will be put as an Enum once I develop the negative grants. I prefer to do it in a separate PR.

File: src/main/java/org/ow2/proactive/catalog/service/BucketGrantService.java [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+
+    /**
+     *
+     * Get the grants assigned to a user and his group
+     *
+     * @param username username
+     * @param userGroup user group
+     * @return all grants that are assigned to a specific username and his group
+     */
+    public List<BucketGrantMetadata> getAllAssignedGrantsForUserAndHisGroup(String username, String userGroup) {
```
Comment (fviale) [inline src/main/java/org/ow2/proactive/catalog/service/BucketGrantService.java:82]: consider having a list or set of groups in parameter and usage of the SQL expression 'in'

File: src/main/java/org/ow2/proactive/catalog/repository/entity/CatalogObjectGrantEntity.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    @GenericGenerator(name = "CATALOG_OBJECT_GRANT_SEQUENCE", strategy = "org.hibernate.id.enhanced.SequenceStyleGenerator", parameters = { @org.hibernate.annotations.Parameter(name = "sequence_name", value = "CATALOG_OBJECT_GRANT_SEQUENCE"),
+                                                                                                                                            @org.hibernate.annotations.Parameter(name = "initial_value", value = "1"),
+                                                                                                                                            @org.hibernate.annotations.Parameter(name = "increment_size", value = "1") })
+    @Column(name = "ID")
+    protected Long id;
+
+    // TODO Change to Enum
+    // Type of this grant: user or group
+    @Column(name = "GRANTEE", nullable = false)
+    protected String grantee;
```
Comment (fviale) [inline src/main/java/org/ow2/proactive/catalog/repository/entity/CatalogObjectGrantEntity.java:60]: rename to granteeType

File: src/main/java/org/ow2/proactive/catalog/repository/entity/CatalogObjectGrantEntity.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+    // The User who created this grant
+    @Column(name = "CREATOR", nullable = false)
+    protected String creator;
+
+    // Username or group Name
+    @Column(name = "PROFITEER", nullable = false)
+    protected String profiteer;
```
Comment (fviale) [inline src/main/java/org/ow2/proactive/catalog/repository/entity/CatalogObjectGrantEntity.java:68]: rename to grantee

File: src/main/java/org/ow2/proactive/catalog/repository/entity/CatalogObjectGrantEntity.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+import lombok.ToString;
+
+
+/**
+ * @author ActiveEon Team
+ */
+@AllArgsConstructor
+@Data
+@Entity
+@Table(name = "CATALOG_OBJECT_GRANT", uniqueConstraints = @UniqueConstraint(columnNames = { "ID" }), indexes = { @Index(name = "ID", columnList = "ID") })
```
Comment (fviale) [inline src/main/java/org/ow2/proactive/catalog/repository/entity/CatalogObjectGrantEntity.java:45]: there should be more indexes here

**Comment by fviale** [comment]
Retest this please

**Review by aminelouati — APPROVED** [review]

**Comment by fviale** [comment]
Retest this please

**Review by fviale — APPROVED** [review]

**Pull request merged**